java/src/main/java/ai/rapids/cudf/HostMemoryBuffer.java

```diff
@@ -156,6 +164,14 @@ public static HostMemoryBuffer allocate(long bytes) {
    * @return the newly created buffer
    */
   public static HostMemoryBuffer allocateRaw(long bytes) {
+    if (ALLOC_DEBUG) {
+      totalAllocated.addAndGet(bytes);
+      double gb = totalAllocated.get() / (double) (1 << 30);
+      if (gb >= lastPrintedGB + 1 || gb <= lastPrintedGB - 1) {
+        lastPrintedGB = gb;
+        log.error("HostMemoryBuffer allocation reaches: " + gb + "GB, stacktrace being: ", new Exception());
```

Comment:
I am confused by exactly why we need the stack trace. I see that the logging can help you get a high water mark on memory usage. But that feature was added to spark-rapids already  https://github.com/NVIDIA/spark-rapids/pull/11725

But what is the stack trace supposed to do? Is it intended to give you a clue on what is allocating all of the memory? If so, then lets actually log all allocations, along with the stack trace (preferably in a machine readable format), and free calls together with the address and the size of the allocation. This would let us truly understand how the memory is being used instead of just sampling a few values that happen to push us over some arbitrary boundary.

Reply:
stacktrace here is supposed to print sampled allocations so that we can roughly know who's using the memory. It happens to be very useful in my recent debug journey. But I agree with you it's not accurate. Will you concern about too many logs if we print "all allocations, along with the stack trace"?

Reply:
I think the recommendation would be still print under a debug flag, so we would get a log back that we can analyze after the fact.

RMM has a similar feature in their tracking allocator (https://github.com/rapidsai/rmm/blob/f9b9f843466b2dcd8872f1707d274ad15be37324/include/rmm/mr/device/tracking_resource_adaptor.hpp#L213) for reference. If our host memory allocator was always a rmm memory resource (right now, the pinned host memory resource is, but not the pageable one if I recall correctly), we could use this to track native stacks (useful if we have native code allocating host memory), but not the java stack. RMM also has a logging adapter that logs all allocations and deallocations for a wrapped resource.

If we are going to log a sample of a stack trace I'd at least add extra locking so we know that the stack trace that caused the max is actually the stack trace we need to focus on. Otherwise it very well could be a false positive.